File: BWChat/Models/DiscoverConfig.swift
import SwiftUI

struct DiscoverConfigData: Codable, Equatable {
    var schemaVersion: Int?
    var sections: [DiscoverSection]
    var meta: DiscoverConfigMeta?

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case sections
        case meta
    }

    var effectiveSections: [DiscoverSection] {
        let normalizedSections = sections
            .filter(\.isEnabled)
            .sorted { $0.sortOrder < $1.sortOrder }
            .compactMap { section in
                var next = section
                next.items = section.items
                    .filter(\.isVisible)
                    .sorted { $0.sortOrder < $1.sortOrder }
                return next.items.isEmpty ? nil : next
            }
        return Self.sectionsPreservingDefaultBlocks(normalizedSections)
    }

    init(
        schemaVersion: Int? = 1,
        sections: [DiscoverSection],
        meta: DiscoverConfigMeta? = nil
    ) {
        self.schemaVersion = schemaVersion
        self.sections = sections
        self.meta = meta
    }

    static let fallback = DiscoverConfigData(sections: defaultSections)

    static let defaultSections: [DiscoverSection] = [
        DiscoverSection(
            id: "social",
            order: 10,
            items: [
                DiscoverItem(
                    id: "moments",
                    titleKey: "discover.moments",
                    systemImage: "camera.fill",
                    colors: ["667EEA", "764BA2"],
                    badgeKey: "moments_unread",
                    dotKey: "moments_new",
                    order: 10,
                    route: DiscoverRoute(type: "native", name: "moments")
                )
            ]
        ),
        DiscoverSection(
            id: "entertainment",
            order: 20,
            items: [
                DiscoverItem(
                    id: "games",
                    titleKey: "discover.games",
                    systemImage: "gamecontroller.fill",
                    colors: ["FF6B6B", "FF8E53"],
                    order: 10,
                    route: DiscoverRoute(type: "native", name: "game_center")
                ),
                DiscoverItem(
                    id: "short_drama",
                    titleKey: "discover.shortDrama",
                    systemImage: "play.rectangle.fill",
                    colors: ["00C6FF", "0072FF"],
                    order: 20,
                    route: DiscoverRoute(type: "native", name: "short_drama")
                ),
                DiscoverItem(
                    id: "live",
                    titleKey: "discover.live",
                    systemImage: "dot.radiowaves.left.and.right",
                    colors: ["FF4D8D", "FF8A3D"],
                    order: 30,
                    route: DiscoverRoute(type: "coming_soon")
                )
            ]
        ),
        DiscoverSection(
            id: "community",
            order: 30,
            items: [
                DiscoverItem(
                    id: "stories",
                    titleKey: "discover.stories",
                    systemImage: "book.closed.fill",
                    colors: ["7F5AF0", "FF7A90"],
                    order: 10,
                    route: DiscoverRoute(type: "native", name: "script_center")
                ),
                DiscoverItem(
                    id: "groups",
                    titleKey: "discover.groups",
                    systemImage: "person.3.fill",
                    colors: ["34C759", "00B894"],
                    order: 20,
                    route: DiscoverRoute(type: "native", name: "groups")
                )
            ]
        ),
        DiscoverSection(
            id: "benefits",
            order: 40,
            items: [
                DiscoverItem(
                    id: "benefits",
                    titleKey: "discover.benefits",
                    systemImage: "gift.fill",
                    colors: ["FFB703", "FB8500"],
                    order: 10,
                    route: DiscoverRoute(type: "coming_soon")
                )
            ]
        )
    ]

    private static let defaultSectionIDs = ["social", "entertainment", "community", "benefits"]

    private static let defaultSectionOrders: [String: Int] = [
        "social": 10,
        "entertainment": 20,
        "community": 30,
        "benefits": 40
    ]

    private static let defaultItemSectionIDs: [String: String] = [
        "moments": "social",
        "games": "entertainment",
        "short_drama": "entertainment",
        "live": "entertainment",
        "stories": "community",
        "groups": "community",
        "benefits": "benefits"
    ]

    private static let movedOutOfDiscoverItemIDs: Set<String> = [
        "nearby",
        "map",
        "map_dating"
    ]

    private static func sectionsPreservingDefaultBlocks(_ sections: [DiscoverSection]) -> [DiscoverSection] {
        var defaultItemsBySectionID: [String: [DiscoverItem]] = [:]
        var customSections: [DiscoverSection] = []

        for section in sections {
            var customItems: [DiscoverItem] = []
            for item in section.items {
                guard !movedOutOfDiscoverItemIDs.contains(item.id.normalizedDiscoverToken) else {
                    continue
                }
                if let sectionID = defaultItemSectionIDs[item.id.normalizedDiscoverToken] {
                    defaultItemsBySectionID[sectionID, default: []].append(item)
                } else {
                    customItems.append(item)
                }
            }

            if !customItems.isEmpty {
                var customSection = section
                customSection.items = customItems.sorted { $0.sortOrder < $1.sortOrder }
                customSections.append(customSection)
            }
        }

        let defaultSections = defaultSectionIDs.compactMap { sectionID -> DiscoverSection? in
            guard let items = defaultItemsBySectionID[sectionID], !items.isEmpty else { return nil }
            let sourceSection = sections.first { $0.id.normalizedDiscoverToken == sectionID }
            return DiscoverSection(
                id: sectionID,
                enabled: true,
                order: sourceSection?.order ?? defaultSectionOrders[sectionID],
                items: items.sorted { $0.sortOrder < $1.sortOrder }
            )
        }

        return (defaultSections + customSections)
            .sorted { $0.sortOrder < $1.sortOrder }
    }
}

struct DiscoverConfigMeta: Codable, Equatable {
    var refreshIntervalSeconds: Int?
    var generatedAt: String?

    enum CodingKeys: String, CodingKey {
        case refreshIntervalSeconds = "refresh_interval_seconds"
        case generatedAt = "generated_at"
    }
}

struct DiscoverSection: Codable, Equatable, Identifiable {
    var id: String
    var enabled: Bool?
    var order: Int?
    var items: [DiscoverItem]

    enum CodingKeys: String, CodingKey {
        case id
        case enabled
        case order
        case items
    }

    init(
        id: String,
        enabled: Bool? = nil,
        order: Int? = nil,
        items: [DiscoverItem]
    ) {
        self.id = id
        self.enabled = enabled
        self.order = order
        self.items = items
    }

    var isEnabled: Bool { enabled ?? true }
    var sortOrder: Int { order ?? 0 }
}

struct DiscoverItem: Codable, Equatable, Identifiable {
    var id: String
    var titleKey: String?
    var title: String?
    var titleI18n: [String: String]?
    var systemImage: String?
    var colors: [String]?
    var badgeKey: String?
    var badgeCount: Int?
    var dotKey: String?
    var showsDot: Bool?
    var enabled: Bool?
    var order: Int?
    var route: DiscoverRoute?

    enum CodingKeys: String, CodingKey {
        case id
        case titleKey = "title_key"
        case title
        case titleI18n = "title_i18n"
        case systemImage = "system_image"
        case colors
        case badgeKey = "badge_key"
        case badgeCount = "badge_count"
        case dotKey = "dot_key"
        case showsDot = "shows_dot"
        case enabled
        case order
        case route
    }

    init(
        id: String,
        titleKey: String? = nil,
        title: String? = nil,
        titleI18n: [String: String]? = nil,
        systemImage: String? = nil,
        colors: [String]? = nil,
        badgeKey: String? = nil,
        badgeCount: Int? = nil,
        dotKey: String? = nil,
        showsDot: Bool? = nil,
        enabled: Bool? = nil,
        order: Int? = nil,
        route: DiscoverRoute? = nil
    ) {
        self.id = id
        self.titleKey = titleKey
        self.title = title
        self.titleI18n = titleI18n
        self.systemImage = systemImage
        self.colors = colors
        self.badgeKey = badgeKey
        self.badgeCount = badgeCount
        self.dotKey = dotKey
        self.showsDot = showsDot
        self.enabled = enabled
        self.order = order
        self.route = route
    }

    var isEnabled: Bool { enabled ?? true }
    var isVisible: Bool { isEnabled && route?.normalizedType != "disabled" }
    var sortOrder: Int { order ?? 0 }

    var displayColors: [Color] {
        let hexValues = (colors ?? [])
            .filter(\.isDiscoverHexColor)
            .prefix(2)
        guard let first = hexValues.first else { return [] }
        guard hexValues.count > 1 else {
            return [Color(hex: first)]
        }
        return hexValues.map(Color.init(hex:))
    }

    func displayTitle(language: AppLanguage) -> String {
        if let preferredTitleKey = preferredLocalTitleKey {
            let localized = L10n.tr(preferredTitleKey)
            if localized != preferredTitleKey { return localized }
        }
        if let localized = titleI18n?.localizedDiscoverValue(for: language) {
            return localized
        }
        if let titleKey, !titleKey.isDiscoverBlank {
            let localized = L10n.tr(titleKey)
            if localized != titleKey { return localized }
        }
        if let title, !title.isDiscoverBlank {
            return title
        }
        return id
    }

    private var preferredLocalTitleKey: String? {
        switch id.normalizedDiscoverToken {
        case "groups", "group", "group_list":
            return "discover.groups"
        case "benefits":
            return "discover.benefits"
        default:
            return nil
        }
    }
}

struct DiscoverRoute: Codable, Equatable {
    var type: String?
    var name: String?
    var url: String?
    var titleKey: String?
    var title: String?
    var titleI18n: [String: String]?
    var messageKey: String?
    var message: String?
    var messageI18n: [String: String]?

    enum CodingKeys: String, CodingKey {
        case type
        case name
        case url
        case titleKey = "title_key"
        case title
        case titleI18n = "title_i18n"
        case messageKey = "message_key"
        case message
        case messageI18n = "message_i18n"
    }

    init(
        type: String? = nil,
        name: String? = nil,
        url: String? = nil,
        titleKey: String? = nil,
        title: String? = nil,
        titleI18n: [String: String]? = nil,
        messageKey: String? = nil,
        message: String? = nil,
        messageI18n: [String: String]? = nil
    ) {
        self.type = type
        self.name = name
        self.url = url
        self.titleKey = titleKey
        self.title = title
        self.titleI18n = titleI18n
        self.messageKey = messageKey
        self.message = message
        self.messageI18n = messageI18n
    }

    var normalizedType: String {
        (type ?? "coming_soon").normalizedDiscoverToken
    }

    func displayTitle(language: AppLanguage, fallback: String) -> String {
        if let localized = titleI18n?.localizedDiscoverValue(for: language) {
            return localized
        }
        if let titleKey, !titleKey.isDiscoverBlank {
            let localized = L10n.tr(titleKey)
            if localized != titleKey { return localized }
        }
        if let title, !title.isDiscoverBlank {
            return title
        }
        return fallback
    }

    func displayMessage(language: AppLanguage) -> String? {
        if let localized = messageI18n?.localizedDiscoverValue(for: language) {
            return localized
        }
        if let messageKey, !messageKey.isDiscoverBlank {
            let localized = L10n.tr(messageKey)
            if localized != messageKey { return localized }
        }
        return message?.isDiscoverBlank == false ? message : nil
    }
}

extension Dictionary where Key == String, Value == String {
    func localizedDiscoverValue(for language: AppLanguage) -> String? {
        let languageCode = language.rawValue
        let candidates = [
            languageCode,
            language.localeIdentifier,
            languageCode.split(separator: "-").first.map(String.init) ?? "",
            "en"
        ]
        for key in candidates {
            if let value = self[key], !value.isDiscoverBlank {
                return value
            }
        }
        return nil
    }
}

extension String {
    var isDiscoverBlank: Bool {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var normalizedDiscoverToken: String {
        trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "-", with: "_")
            .lowercased()
    }

    var isDiscoverHexColor: Bool {
        let value = trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        guard value.count == 6 || value.count == 8 else { return false }
        return UInt64(value, radix: 16) != nil
    }
}

File: expo-app/src/services/discover/DiscoverConfig.ts
import type { DynamicRoute } from "@/services/remote-config/types";

export interface DiscoverConfigMeta {
  refreshIntervalSeconds?: number | undefined;
  generatedAt?: string | undefined;
}

export interface DiscoverItem {
  id: string;
  titleKey?: string | undefined;
  title?: string | undefined;
  titleI18n?: Record<string, string> | undefined;
  systemImage?: string | undefined;
  colors?: string[] | undefined;
  badgeKey?: string | undefined;
  badgeCount?: number | undefined;
  dotKey?: string | undefined;
  showsDot?: boolean | undefined;
  enabled?: boolean | undefined;
  order?: number | undefined;
  route?: DynamicRoute | undefined;
}

export interface DiscoverSection {
  id: string;
  enabled?: boolean | undefined;
  order?: number | undefined;
  items: DiscoverItem[];
}

export interface DiscoverConfigData {
  schemaVersion?: number | undefined;
  sections: DiscoverSection[];
  meta?: DiscoverConfigMeta | undefined;
}

export const defaultDiscoverSections: DiscoverSection[] = [
  {
    id: "social",
    order: 10,
    items: [{
      id: "moments",
      titleKey: "discover.moments",
      systemImage: "camera.fill",
      colors: ["667EEA", "764BA2"],
      badgeKey: "moments_unread",
      dotKey: "moments_new",
      order: 10,
      route: { type: "native", name: "moments" },
    }],
  },
  {
    id: "entertainment",
    order: 20,
    items: [
      { id: "games", titleKey: "discover.games", systemImage: "gamecontroller.fill", colors: ["FF6B6B", "FF8E53"], order: 10, route: { type: "native", name: "game_center" } },
      { id: "stories", titleKey: "discover.stories", systemImage: "book.closed.fill", colors: ["7F5AF0", "FF7A90"], order: 20, route: { type: "native", name: "script_center" } },
      { id: "short_drama", titleKey: "discover.shortDrama", systemImage: "play.rectangle.fill", colors: ["00C6FF", "0072FF"], order: 30, route: { type: "native", name: "short_drama" } },
      { id: "live", titleKey: "discover.live", systemImage: "video.fill", colors: ["FF4D8D", "FF8A3D"], order: 40, route: { type: "coming_soon" } },
    ],
  },
  {
    id: "community",
    order: 30,
    items: [{ id: "groups", titleKey: "discover.groups", systemImage: "person.3.fill", colors: ["34C759", "00B894"], order: 10, route: { type: "native", name: "groups" } }],
  },
  {
    id: "benefits",
    order: 40,
    items: [{ id: "benefits", titleKey: "discover.benefits", systemImage: "gift.fill", colors: ["FFB703", "FB8500"], order: 10, route: { type: "native", name: "activity_center" } }],
  },
];

const defaultSectionIds = ["social", "entertainment", "community", "benefits"] as const;
const defaultSectionOrders: Record<string, number> = { social: 10, entertainment: 20, community: 30, benefits: 40 };
const defaultItemSectionIds: Record<string, string> = {
  moments: "social",
  games: "entertainment",
  stories: "entertainment",
  short_drama: "entertainment",
  live: "entertainment",
  groups: "community",
  benefits: "benefits",
};
const defaultItemOrders: Record<string, number> = { moments: 10, games: 10, stories: 20, short_drama: 30, live: 40, groups: 10, benefits: 10 };
const movedOutOfDiscover = new Set(["nearby", "map", "map_dating"]);

export function parseDiscoverConfig(value: unknown): DiscoverConfigData {
  const root = unwrapDiscover(value);
  const sections = arrayValue(root.sections).map(normalizeSection).filter(notNull);
  return {
    ...(numberValue(root.schema_version, root.schemaVersion) !== undefined
      ? { schemaVersion: numberValue(root.schema_version, root.schemaVersion) }
      : {}),
    sections,
    ...(isRecord(root.meta) ? {
      meta: {
        ...(numberValue(root.meta.refresh_interval_seconds, root.meta.refreshIntervalSeconds) !== undefined
          ? { refreshIntervalSeconds: numberValue(root.meta.refresh_interval_seconds, root.meta.refreshIntervalSeconds) }
          : {}),
        ...(stringValue(root.meta.generated_at, root.meta.generatedAt)
          ? { generatedAt: stringValue(root.meta.generated_at, root.meta.generatedAt) }
          : {}),
      },
    } : {}),
  };
}

export function effectiveDiscoverSections(config: DiscoverConfigData): DiscoverSection[] {
  const normalized = config.sections
    .filter((section) => section.enabled !== false)
    .sort(byOrder)
    .map((section) => ({
      ...section,
      items: section.items
        .filter((item) => item.enabled !== false && normalizeToken(item.route?.type ?? "coming_soon") !== "disabled")
        .sort(byOrder),
    }))
    .filter((section) => section.items.length > 0);
  return preserveDefaultBlocks(normalized);
}

export function discoverItemTitle(
  item: DiscoverItem,
  language: string,
  translate: (key: string) => string,
): string {
  const preferred = preferredTitleKey(item.id);
  if (preferred) {
    const localized = translate(preferred);
    if (localized !== preferred) return localized;
  }
  const translated = localizedValue(item.titleI18n, language);
  if (translated) return translated;
  const titleKey = item.titleKey?.trim();
  if (titleKey) {
    const localized = translate(titleKey);
    if (localized !== titleKey) return localized;
  }
  return item.title?.trim() || item.id;
}

export function normalizeDiscoverColor(value: string | undefined): string | undefined {
  if (!value) return undefined;
  const cleaned = value.trim().replace(/[^a-fA-F0-9]/g, "");
  if ((cleaned.length !== 6 && cleaned.length !== 8) || !/^[a-fA-F0-9]+$/.test(cleaned)) return undefined;
  return `#${cleaned}`;
}

export function normalizeToken(value: string): string {
  return value.trim().replaceAll("-", "_").toLocaleLowerCase();
}

function preserveDefaultBlocks(sections: DiscoverSection[]): DiscoverSection[] {
  const defaultItems = new Map<string, DiscoverItem[]>();
  const customSections: DiscoverSection[] = [];
  for (const section of sections) {
    const custom: DiscoverItem[] = [];
    for (const item of section.items) {
      const id = normalizeToken(item.id);
      if (movedOutOfDiscover.has(id)) continue;
      const defaultSection = defaultItemSectionIds[id];
      if (defaultSection) {
        const next = { ...item, order: defaultItemOrders[id] ?? item.order };
        defaultItems.set(defaultSection, [...(defaultItems.get(defaultSection) ?? []), next]);
      } else {
        custom.push(item);
      }
    }
    if (custom.length > 0) customSections.push({ ...section, items: custom.sort(byOrder) });
  }
  const stableSections = defaultSectionIds.flatMap((id) => {
    const items = defaultItems.get(id);
    if (!items?.length) return [];
    const source = sections.find((section) => normalizeToken(section.id) === id);
    return [{ id, enabled: true, order: source?.order ?? defaultSectionOrders[id], items: items.sort(byOrder) }];
  });
  return [...stableSections, ...customSections].sort(byOrder);
}

function normalizeSection(value: unknown): DiscoverSection | null {
  if (!isRecord(value)) return null;
  const id = stringValue(value.id);
  if (!id) return null;
  return {
    id,
    ...(boolValue(value.enabled) !== undefined ? { enabled: boolValue(value.enabled) } : {}),
    ...(numberValue(value.order) !== undefined ? { order: numberValue(value.order) } : {}),
    items: arrayValue(value.items).map(normalizeItem).filter(notNull),
  };
}

function normalizeItem(value: unknown): DiscoverItem | null {
  if (!isRecord(value)) return null;
  const id = stringValue(value.id);
  if (!id) return null;
  return {
    id,
    ...optionalString("titleKey", value.title_key, value.titleKey),
    ...optionalString("title", value.title),
    ...optionalStringMap("titleI18n", value.title_i18n, value.titleI18n),
    ...optionalString("systemImage", value.system_image, value.systemImage),
    ...(arrayValue(value.colors).map((color) => stringValue(color)).filter(notNull).length > 0
      ? { colors: arrayValue(value.colors).map((color) => stringValue(color)).filter(notNull) }
      : {}),
    ...optionalString("badgeKey", value.badge_key, value.badgeKey),
    ...(numberValue(value.badge_count, value.badgeCount) !== undefined ? { badgeCount: numberValue(value.badge_count, value.badgeCount) } : {}),
    ...optionalString("dotKey", value.dot_key, value.dotKey),
    ...(boolValue(value.shows_dot, value.showsDot) !== undefined ? { showsDot: boolValue(value.shows_dot, value.showsDot) } : {}),
    ...(boolValue(value.enabled) !== undefined ? { enabled: boolValue(value.enabled) } : {}),
    ...(numberValue(value.order) !== undefined ? { order: numberValue(value.order) } : {}),
    ...(isRecord(value.route) ? { route: normalizeRoute(value.route) } : {}),
  };
}

function normalizeRoute(value: Record<string, unknown>): DynamicRoute {
  return {
    ...optionalString("type", value.type),
    ...optionalString("name", value.name),
    ...optionalString("url", value.url),
    ...optionalString("screenId", value.screen_id, value.screenId),
    ...optionalString("titleKey", value.title_key, value.titleKey),
    ...optionalString("title", value.title),
    ...optionalStringMap("titleI18n", value.title_i18n, value.titleI18n),
    ...optionalString("messageKey", value.message_key, value.messageKey),
    ...optionalString("message", value.message),
    ...optionalStringMap("messageI18n", value.message_i18n, value.messageI18n),
    ...(isRecord(value.params) ? { params: value.params } : {}),
  };
}

function unwrapDiscover(value: unknown): Record<string, unknown> {
  if (!isRecord(value)) throw new Error("Discover config payload is invalid");
  if (isRecord(value.data)) return unwrapDiscover(value.data);
  if (isRecord(value.config)) return unwrapDiscover(value.config);
  return value;
}

function localizedValue(values: Record<string, string> | undefined, language: string): string | undefined {
  if (!values) return undefined;
  const normalized = language.replaceAll("_", "-");
  const base = normalized.split("-")[0] ?? "";
  for (const key of [language, normalized, base, "en"]) {
    const value = values[key]?.trim();
    if (value) return value;
  }
  return undefined;
}

function preferredTitleKey(id: string): string | undefined {
  switch (normalizeToken(id)) {
    case "live": return "discover.live";
    case "groups":
    case "group":
    case "group_list": return "discover.groups";
    case "benefits": return "discover.benefits";
    default: return undefined;
  }
}

function byOrder<T extends { order?: number | undefined }>(left: T, right: T): number {
  return (left.order ?? 0) - (right.order ?? 0);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function arrayValue(value: unknown): unknown[] { return Array.isArray(value) ? value : []; }
function stringValue(...values: unknown[]): string | undefined {
  for (const value of values) if (typeof value === "string" && value.trim()) return value.trim();
  return undefined;
}
function numberValue(...values: unknown[]): number | undefined {
  for (const value of values) {
    const number = typeof value === "number" ? value : typeof value === "string" ? Number(value) : Number.NaN;
    if (Number.isFinite(number)) return number;
  }
  return undefined;
}
function boolValue(...values: unknown[]): boolean | undefined {
  for (const value of values) if (typeof value === "boolean") return value;
  return undefined;
}
function optionalString<Key extends string>(key: Key, ...values: unknown[]): Partial<Record<Key, string>> {
  const value = stringValue(...values);
  return value ? { [key]: value } as Partial<Record<Key, string>> : {};
}
function optionalStringMap<Key extends string>(key: Key, ...values: unknown[]): Partial<Record<Key, Record<string, string>>> {
  const source = values.find(isRecord);
  if (!source) return {};
  const result = Object.fromEntries(Object.entries(source).flatMap(([entryKey, entryValue]) => {
    const text = stringValue(entryValue);
    return text ? [[entryKey, text]] : [];
  }));
  return Object.keys(result).length > 0 ? { [key]: result } as Partial<Record<Key, Record<string, string>>> : {};
}
function notNull<T>(value: T | null | undefined): value is T { return value !== null && value !== undefined; }
